src/core/tls.go

```diff
@@ -103,11 +103,14 @@ func (t *tcptls) configForOptions(options *tcpOptions) *tls.Config {
 		}
 		return nil
 	}
-	return &config
+	if serverName != "" {
+		config.ServerName = serverName
```

Comment:
See above, it seems like `config.ServerName = tcpOptions.tlsSNI` would unconditionally do the right thing?